core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -126,6 +125,31 @@ class KafkaApis(val requestChannel: RequestChannel,
     info("Shutdown complete.")
   }
 
+  private def maybeForward(request: RequestChannel.Request,
+                           handler: RequestChannel.Request => Unit): Unit = {
+    if (request.envelopeContext.isDefined && request.principalSerde.isEmpty) {
+      sendErrorResponseMaybeThrottle(request, Errors.PRINCIPAL_DESERIALIZATION_FAILURE.exception())
+    } else if (request.envelopeContext.isDefined &&
+      (!request.context.fromPrivilegedListener ||
+      !authorize(request.envelopeContext.get.brokerContext, CLUSTER_ACTION, CLUSTER, CLUSTER_NAME))
+    ) {
+      // If the designated forwarding request is not coming from a privileged listener, or
+      // it fails CLUSTER_ACTION permission, we would fail the authorization.
+      sendErrorResponseMaybeThrottle(request, Errors.BROKER_AUTHORIZATION_FAILURE.exception())
+    } else if (request.envelopeContext.isDefined && !controller.isActive) {
+      sendErrorResponseMaybeThrottle(request, Errors.NOT_CONTROLLER.exception())
+    } else if (!controller.isActive && couldDoRedirection(request)) {
+      redirectionManager.forwardRequest(sendResponseMaybeThrottle, request)
+    } else {
+      // When IBP is smaller than 2.8 or the principal serde is undefined, forwarding is not supported,
+      // therefore requests are handled directly.
+      handler(request)
+    }
+  }
+
+  private def couldDoRedirection(request: RequestChannel.Request): Boolean =
```

Comment:
We use 'forward' and 'redirect' interchangeably throughout the PR, but the names do suggest different behavior. In my mind 'redirection' suggests that we are telling the client to go somewhere else, while 'forward' suggests that the broker is passing the request through to its destination. So maybe we can stick with 'forward' consistently (e.g. `isForwardingEnabled`)?

Reply:
Sounds good.